# 频繁项挖掘

In [1]:
import numpy as np
import pandas as pd
import itertools

In [2]:
# 创建一个简单的数据集userBuyItems, 其中每一项表示用户购买的商品组合
userBuyItems=[[1, 3, 4],[2, 3, 5],[1, 2, 3, 5],[2, 5]]
userBuyItems

[[1, 3, 4], [2, 3, 5], [1, 2, 3, 5], [2, 5]]

In [3]:
# 获取商品的集合 goods 

tmp=[]
for d1 in userBuyItems:
    for item in d1:
        tmp.append(item)
goods=np.unique(tmp)
goods

array([1, 2, 3, 4, 5])

In [4]:
# 计算商品在购买组合中出现的频率FREQ1(次数)
FREQ1_TEMP={}

# 初始化
for good in goods:
    FREQ1_TEMP[good]=0
    
# 计算单个商品的频率
for userBuyItem in userBuyItems:
    for good in goods:
        if good in userBuyItem:
            FREQ1_TEMP[good]+=1
            print("good ",good,"is in ",userBuyItem)
        else:
            print("=======good ",good,"is not in ",userBuyItem)
FREQ1=np.array([i for i in FREQ1_TEMP.items()])
FREQ1

good  1 is in  [1, 3, 4]
=======good  2 is not in  [1, 3, 4]
good  3 is in  [1, 3, 4]
good  4 is in  [1, 3, 4]
=======good  5 is not in  [1, 3, 4]
=======good  1 is not in  [2, 3, 5]
good  2 is in  [2, 3, 5]
good  3 is in  [2, 3, 5]
=======good  4 is not in  [2, 3, 5]
good  5 is in  [2, 3, 5]
good  1 is in  [1, 2, 3, 5]
good  2 is in  [1, 2, 3, 5]
good  3 is in  [1, 2, 3, 5]
=======good  4 is not in  [1, 2, 3, 5]
good  5 is in  [1, 2, 3, 5]
=======good  1 is not in  [2, 5]
good  2 is in  [2, 5]
=======good  3 is not in  [2, 5]
=======good  4 is not in  [2, 5]
good  5 is in  [2, 5]


array([[1, 2],
       [2, 3],
       [3, 3],
       [4, 1],
       [5, 3]])

In [5]:
# 计算单个商品的支持度 SP1
SP1=np.divide(FREQ1,np.array([1,len(goods)]))
SP1

array([[1. , 0.4],
       [2. , 0.6],
       [3. , 0.6],
       [4. , 0.2],
       [5. , 0.6]])

In [6]:
# 获取 支持度>0.5 的商品
minSupport=0
SUPPORT1_R1=SP1[SP1[:,1]>minSupport]

In [7]:
# 生成二元组
BINARY_SUPPORT_TEMP=[]
for item in itertools.combinations(SUPPORT1_R1[:,0],2):
    BINARY_SUPPORT_TEMP.append(np.array(item))

BINARY_ITEMS=np.array(BINARY_SUPPORT_TEMP)
BINARY_ITEMS

array([[1., 2.],
       [1., 3.],
       [1., 4.],
       [1., 5.],
       [2., 3.],
       [2., 4.],
       [2., 5.],
       [3., 4.],
       [3., 5.],
       [4., 5.]])

In [8]:
# 计算二元组的支持度
# 计算单个商品的频率
FREQ2={}
for good_conb_binary in BINARY_ITEMS:
    FREQ2[str(np.array(good_conb_binary,dtype=np.int32))]=0
print(FREQ2)
for userBuyItem in userBuyItems:
    for good_conb_binary in BINARY_ITEMS:
        good_conb_binary=np.array(good_conb_binary,dtype=np.int32)
        userBuyItem=np.array(userBuyItem,dtype=np.int32)
        intersect=np.intersect1d(good_conb_binary,userBuyItem)
        # 处理空集
        if len(intersect)==0:
            intersect=np.array([1,1])
            
        # 计算出现的频率
        print("intersect",intersect,"good_conb ",good_conb_binary,"userBuyItem",userBuyItem, (intersect==good_conb_binary).all())
        if (intersect==good_conb_binary).all():
            FREQ2[str(good_conb_binary)]+=1
#         if  in userBuyItem:
#             FREQ1_TEMP[good]+=1
#             print("good_conb ",good_conb,"is in ",userBuyItem)
#         else:
#             print("=======good ",good_conb,"is not in ",userBuyItem)
FREQ2

{'[1 2]': 0, '[1 3]': 0, '[1 4]': 0, '[1 5]': 0, '[2 3]': 0, '[2 4]': 0, '[2 5]': 0, '[3 4]': 0, '[3 5]': 0, '[4 5]': 0}
intersect [1] good_conb  [1 2] userBuyItem [1 3 4] False
intersect [1 3] good_conb  [1 3] userBuyItem [1 3 4] True
intersect [1 4] good_conb  [1 4] userBuyItem [1 3 4] True
intersect [1] good_conb  [1 5] userBuyItem [1 3 4] False
intersect [3] good_conb  [2 3] userBuyItem [1 3 4] False
intersect [4] good_conb  [2 4] userBuyItem [1 3 4] False
intersect [1 1] good_conb  [2 5] userBuyItem [1 3 4] False
intersect [3 4] good_conb  [3 4] userBuyItem [1 3 4] True
intersect [3] good_conb  [3 5] userBuyItem [1 3 4] False
intersect [4] good_conb  [4 5] userBuyItem [1 3 4] False
intersect [2] good_conb  [1 2] userBuyItem [2 3 5] False
intersect [3] good_conb  [1 3] userBuyItem [2 3 5] False
intersect [1 1] good_conb  [1 4] userBuyItem [2 3 5] False
intersect [5] good_conb  [1 5] userBuyItem [2 3 5] False
intersect [2 3] good_conb  [2 3] userBuyItem [2 3 5] True
intersect [2] go

{'[1 2]': 1,
 '[1 3]': 2,
 '[1 4]': 1,
 '[1 5]': 1,
 '[2 3]': 2,
 '[2 4]': 0,
 '[2 5]': 3,
 '[3 4]': 1,
 '[3 5]': 2,
 '[4 5]': 0}

In [9]:
SP2={}
for i,v in FREQ2.items():
    print(i,v)
    SP2[i]=np.divide(v,len(goods))
SP2

[1 2] 1
[1 3] 2
[1 4] 1
[1 5] 1
[2 3] 2
[2 4] 0
[2 5] 3
[3 4] 1
[3 5] 2
[4 5] 0


{'[1 2]': 0.2,
 '[1 3]': 0.4,
 '[1 4]': 0.2,
 '[1 5]': 0.2,
 '[2 3]': 0.4,
 '[2 4]': 0.0,
 '[2 5]': 0.6,
 '[3 4]': 0.2,
 '[3 5]': 0.4,
 '[4 5]': 0.0}

In [10]:
def getFirstItemSupport(key):
    return (SP1[SP1[:,0]==float(key[1])][0][1])
    
# 计算二元组的可信度
for key,val in SP2.items():
    print( key[1],"-->",key[3],":",val/getFirstItemSupport(key))
print("原始项目",userBuyItems)

1 --> 2 : 0.5
1 --> 3 : 1.0
1 --> 4 : 0.5
1 --> 5 : 0.5
2 --> 3 : 0.6666666666666667
2 --> 4 : 0.0
2 --> 5 : 1.0
3 --> 4 : 0.33333333333333337
3 --> 5 : 0.6666666666666667
4 --> 5 : 0.0
原始项目 [[1, 3, 4], [2, 3, 5], [1, 2, 3, 5], [2, 5]]


购买2之后，购买3的概率是0.67
购买3只后，购买5的概率是1